### PR TITLE
source-pcap-file: Handle Pcap File Init Failing (#1694)

### PR DESCRIPTION
**Pull request by dbcfd**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
- Better handle failure in pcap file init

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

**Comment by dbcfd** [comment]
Still seeing the engine fail to properly handle this.

`[4355] 18/2/2019 -- 09:53:53 - (source-pcap-file.c:271) <Warning> (ReceivePcapFileThreadInit) -- [ERRCODE: SC_ERR_PCAP_DISPATCH(20)] - Failed to init pcap file empty.pcap, skipping
[4355] 18/2/2019 -- 09:53:53 - (runmode-unix-socket.c:616) <Info> (UnixSocketPcapFile) -- Marking current task as failed
[4611] 18/2/2019 -- 09:53:53 - (tm-threads.c:2127) <Error> (TmThreadWaitOnThreadInit) -- [ERRCODE: SC_ERR_THREAD_INIT(49)] - thread "RX#01" failed to initialize: flags 0147
[4611] 18/2/2019 -- 09:53:53 - (runmode-unix-socket.c:473) <Info> (UnixSocketPcapFilesCheck) -- Preceeding task failed, cleaning the running mode
[4611] 18/2/2019 -- 09:53:53 - (runmode-unix-socket.c:478) <Info> (UnixSocketPcapFilesCheck) -- Resetting engine state
[4611] 18/2/2019 -- 09:54:54 - (tm-threads.c:1578) <Error> (TmThreadDisableReceiveThreads) -- [ERRCODE: SC_ERR_FATAL(171)] - Engine unable to disable detect thread - "RX#01". Killing engine`

**Comment by victorjulien** [comment]
Can you paste the output of a unix run with a bad pcap?

**Comment by dbcfd** [comment]
Here's the output after changes: `[4099] 18/2/2019 -- 16:09:14 - (source-pcap-file.c:243) <Info> (ReceivePcapFileThreadInit) -- Checking file or directory empty.pcap
[4099] 18/2/2019 -- 16:09:14 - (source-pcap-file-directory-helper.c:212) <Info> (PcapDetermineDirectoryOrFile) -- empty.pcap: Plain file, not a directory
[4099] 18/2/2019 -- 16:09:14 - (source-pcap-file.c:251) <Info> (ReceivePcapFileThreadInit) -- Argument empty.pcap was a file
[4099] 18/2/2019 -- 16:09:14 - (source-pcap-file-helper.c:174) <Error> (InitPcapFile) -- [ERRCODE: SC_ERR_FOPEN(44)] - truncated dump file; tried to read 4 file header bytes, only got 0
[4099] 18/2/2019 -- 16:09:14 - (source-pcap-file.c:278) <Warning> (ReceivePcapFileThreadInit) -- [ERRCODE: SC_ERR_PCAP_DISPATCH(20)] - Failed to init pcap file empty.pcap, skipping
[4355] 18/2/2019 -- 16:09:14 - (tm-threads.c:2172) <Notice> (TmThreadWaitOnThreadInit) -- all 12 packet processing threads, 5 management threads initialized, engine started.
[4099] 18/2/2019 -- 16:09:14 - (source-pcap-file.c:160) <Error> (ReceivePcapFileLoop) -- [ERRCODE: SC_ERR_INVALID_ARGUMENT(13)] - error: data == NULL
[4099] 18/2/2019 -- 16:09:14 - (runmode-unix-socket.c:616) <Info> (UnixSocketPcapFile) -- Marking current task as failed
[4355] 18/2/2019 -- 16:09:14 - (runmode-unix-socket.c:473) <Info> (UnixSocketPcapFilesCheck) -- Preceeding task failed, cleaning the running mode
[4355] 18/2/2019 -- 16:09:14 - (runmode-unix-socket.c:478) <Info> (UnixSocketPcapFilesCheck) -- Resetting engine state
[4355] 18/2/2019 -- 16:09:15 - (counters.c:849) <Info> (StatsLogSummary) -- Alerts: 0
[4355] 18/2/2019 -- 16:09:24 - (runmode-unix-socket.c:413) <Info> (UnixSocketAddPcapFileImpl) -- Added file 'empty.pcap' to list
[4355] 18/2/2019 -- 16:09:24 - (runmode-unix-socket.c:558) <Info> (UnixSocketPcapFilesCheck) -- pcap-file.tenant-id not set
[4355] 18/2/2019 -- 16:09:24 - (runmode-unix-socket.c:572) <Info> (UnixSocketPcapFilesCheck) -- Starting run for 'empty.pcap'
[4355] 18/2/2019 -- 16:09:24 - (util-logopenfile.c:501) <Info> (SCConfLogOpenGeneric) -- eve-log output device (regular) initialized: suricata.alerts
[4355] 18/2/2019 -- 16:09:24 - (output-json-stats.c:468) <Warning> (OutputStatsLogInitSub) -- [ERRCODE: SC_WARN_EVE_MISSING_EVENTS(318)] - eve.stats will not display all decoder events correctly. See #2225. Set a prefix in stats.decoder-events-prefix. In 5.0 the prefix will default to 'decoder.event'.
[10255] 18/2/2019 -- 16:09:24 - (source-pcap-file.c:243) <Info> (ReceivePcapFileThreadInit) -- Checking file or directory empty.pcap
[10255] 18/2/2019 -- 16:09:24 - (source-pcap-file-directory-helper.c:212) <Info> (PcapDetermineDirectoryOrFile) -- empty.pcap: Plain file, not a directory
[10255] 18/2/2019 -- 16:09:24 - (source-pcap-file.c:251) <Info> (ReceivePcapFileThreadInit) -- Argument empty.pcap was a file
[10255] 18/2/2019 -- 16:09:24 - (source-pcap-file-helper.c:174) <Error> (InitPcapFile) -- [ERRCODE: SC_ERR_FOPEN(44)] - truncated dump file; tried to read 4 file header bytes, only got 0
[10255] 18/2/2019 -- 16:09:24 - (source-pcap-file.c:278) <Warning> (ReceivePcapFileThreadInit) -- [ERRCODE: SC_ERR_PCAP_DISPATCH(20)] - Failed to init pcap file empty.pcap, skipping
[4355] 18/2/2019 -- 16:09:24 - (tm-threads.c:2172) <Notice> (TmThreadWaitOnThreadInit) -- all 12 packet processing threads, 5 management threads initialized, engine started.
[10255] 18/2/2019 -- 16:09:24 - (source-pcap-file.c:160) <Error> (ReceivePcapFileLoop) -- [ERRCODE: SC_ERR_INVALID_ARGUMENT(13)] - error: data == NULL
[10255] 18/2/2019 -- 16:09:24 - (runmode-unix-socket.c:616) <Info> (UnixSocketPcapFile) -- Marking current task as failed
[4355] 18/2/2019 -- 16:09:24 - (runmode-unix-socket.c:473) <Info> (UnixSocketPcapFilesCheck) -- Preceeding task failed, cleaning the running mode
[4355] 18/2/2019 -- 16:09:24 - (runmode-unix-socket.c:478) <Info> (UnixSocketPcapFilesCheck) -- Resetting engine state
[4355] 18/2/2019 -- 16:09:25 - (counters.c:849) <Info> (StatsLogSummary) -- Alerts: 0
[4355] 18/2/2019 -- 16:09:35 - (unix-manager.c:574) <Info> (UnixCommandRun) -- Unix socket: lost connection with client
^C[775] 18/2/2019 -- 16:09:36 - (suricata.c:2851) <Notice> (SuricataMainLoop) -- Signal Received.  Stopping engine.
[775] 18/2/2019 -- 16:09:36 - (detect-engine-build.c:1733) <Info> (SigAddressCleanupStage1) -- cleaning up signature grouping structure... complete`

**Comment by victorjulien** [comment]
Think it would be good to remove this:
```
[10255] 18/2/2019 -- 16:09:24 - (source-pcap-file.c:160) <Error> (ReceivePcapFileLoop) -- [ERRCODE: SC_ERR_INVALID_ARGUMENT(13)] - error: data == NULL
```
or replace it with something that will have meaning to a user.

File: src/source-pcap-file.c [hunk 1]
```diff
@@ -156,6 +157,14 @@ TmEcode ReceivePcapFileLoop(ThreadVars *tv, void *data, void *slot)
 {
     SCEnter();
 
+    if(unlikely(data == NULL)) {
+        SCLogError(SC_ERR_INVALID_ARGUMENT, "ReceivePcapFileLoop: data == NULL, this may indicate that thread init was incomplete when running as unix socket");
```
Comment (victorjulien) [inline src/source-pcap-file.c]: what about something like: "pcap reader thread failed to initialize"

**Comment by dbcfd** [comment]
Closing for https://github.com/OISF/suricata/pull/3672

**Pull request closed without merging**